src/main/java/com/google/devtools/build/lib/remote/blobstore/s3/S3BlobStore.java

```diff
@@ -0,0 +1,159 @@
+package com.google.devtools.build.lib.remote.blobstore.s3;
+
+import com.amazonaws.ClientConfiguration;
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.auth.AWSCredentialsProvider;
+import com.amazonaws.services.s3.AmazonS3ClientBuilder;
+import com.amazonaws.services.s3.AmazonS3URI;
+import com.amazonaws.services.s3.model.ObjectMetadata;
+import com.amazonaws.services.s3.model.S3Object;
+import com.google.common.base.Joiner;
+import com.google.common.io.ByteStreams;
+import com.google.common.io.Closeables;
+import com.google.devtools.build.lib.remote.blobstore.SimpleBlobStore;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+import static com.amazonaws.Protocol.HTTPS;
+import static com.amazonaws.regions.Regions.DEFAULT_REGION;
+import static com.google.devtools.build.lib.remote.blobstore.s3.CacheType.*;
+
+/**
+ * Implementation of {@link SimpleBlobStore} that can talk directly to an S3 bucket backend.
+ *
+ * <p>Blobs (Binary large objects) are uploaded using the {@code putObject} operation.
+ * Action cache blobs are stored under the path {@code /ac/base16-key}.
+ * CAS (Content Addressable Storage) blobs are stored under the path {@code /cas/base16-key}.
+ *
+ * <p>Blobs are downloaded using the {@code getObject} operation at the paths they were stored at.
+ *
+ * <p>This blobstore always uses TLS.
+ */
+public class S3BlobStore implements SimpleBlobStore {
```

Comment:
I was hoping that we could reuse the HttpBlobStore, by using the AWS S3 HTTP API

https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectGET.html
https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPUT.html

Do you think this would be possible?